### DISCUSSION TITLE
restitution not working for resting sphere

### DISCUSSION THREAD

**Discussion by wedesoft** [jrouwe/JoltPhysics#1407] [Q&A]
Hi, thanks for developing JoltPhysics and releasing it as open source!
I noticed a peculiar issue when testing a collision of a moving and a resting sphere.
If two spheres are colliding head on with restitution 1.0, I would expect them to exchange speeds.
If the spheres both have non-zero speed in collision direction it works.
However when [testing collision with a resting sphere](https://github.com/wedesoft/sfsim/blob/3dd3dcc40f237633e8eb1062e9ecfd2508de60b9/test/clj/sfsim/t_jolt.clj#L93), both spheres end up sticking together and moving with the average initial speed:

```
sphere1 =  #vec3 [0.0, -1.0, 0.0] , sphere2 =  #vec3 [0.0, 0.0, 0.0]
sphere1 =  #vec3 [0.0, -1.0, 0.0] , sphere2 =  #vec3 [0.0, 0.0, 0.0]
...
sphere1 =  #vec3 [0.0, -1.0, 0.0] , sphere2 =  #vec3 [0.0, 0.0, 0.0]
sphere1 =  #vec3 [0.0, -1.0, 0.0] , sphere2 =  #vec3 [0.0, 0.0, 0.0]
sphere1 =  #vec3 [0.0, -1.0, 0.0] , sphere2 =  #vec3 [0.0, 0.0, 0.0]
sphere1 =  #vec3 [0.0, -0.5, 0.0] , sphere2 =  #vec3 [0.0, -0.5, 0.0]
sphere1 =  #vec3 [0.0, -0.5, 0.0] , sphere2 =  #vec3 [0.0, -0.5, 0.0]
...
sphere1 =  #vec3 [0.0, -0.5, 0.0] , sphere2 =  #vec3 [0.0, -0.5, 0.0]
sphere1 =  #vec3 [0.0, -0.5, 0.0] , sphere2 =  #vec3 [0.0, -0.5, 0.0]
sphere1 =  #vec3 [0.0, -0.5, 0.0] , sphere2 =  #vec3 [0.0, -0.5, 0.0]
```

I am using Clojure and C++. Here are the [C++ methods for accessing Jolt](https://github.com/wedesoft/sfsim/blob/3dd3dcc40f237633e8eb1062e9ecfd2508de60b9/src/c/sfsim/jolt.cc#L203).
The code for the test is as follows:

```Clojure
  (fact "Test restitution of sphere"
        (let [sphere1 (make-sphere 1.0 1000.0 (vec3 0.0  2.0 0.0) (q/->Quaternion 1 0 0 0) (vec3 0 -1 0) (vec3 0 0 0))
              sphere2 (make-sphere 1.0 1000.0 (vec3 0.0 -2.0 0.0) (q/->Quaternion 1 0 0 0) (vec3 0  0 0) (vec3 0 0 0))]
          (set-gravity (vec3 0 0 0))
          (set-friction sphere1 0.2)
          (set-restitution sphere1 1.0)
          (set-friction sphere2 0.2)
          (set-restitution sphere2 1.0)
          (optimize-broad-phase)
          (dotimes [i 40]
            (activate-body sphere1)
            (activate-body sphere2)
            (update-system 0.1 1)
            (println "sphere1 = " (get-linear-velocity sphere1) ", sphere2 = " (get-linear-velocity sphere2)))
          (remove-and-destroy-body sphere1)
          (remove-and-destroy-body sphere2)))
```

The issue persists even when setting gravity to [0, 0, 1]. I.e. the sticking behaviour occurs when the y-component of the speed of the second sphere is 0.
Kind regards
Jan

**Comment by wedesoft** [comment]
I can try whether compiler options make a difference. At the moment I use this configuration:
```Shell
./cmake_linux_clang_gcc.sh Release g++ -DCMAKE_POSITION_INDEPENDENT_CODE=ON -DDOUBLE_PRECISION=ON -DDEBUG_RENDERER_IN_DEBUG_AND_RELEASE=OFF -DPROFILER_IN_DEBUG_AND_RELEASE=OFF -DUSE_AVX2=OFF -DUSE_LZCNT=OFF -DUSE_TZCNT=OFF -DUSE_F16C=OFF -DUSE_FMADD=OFF
```

**Comment by wedesoft** [comment]
~~Hi, I have created a [branch with failing test](https://github.com/jrouwe/JoltPhysics/compare/master...wedesoft:JoltPhysics:restitution-with-resting-object) in case this helps.~~ (ok, deleted because issue resolved)

**Comment by jrouwe** [comment]
The velocity of (0, 1, 0) happens to be exactly the default minimum velocity needed for an elastic collision to happen:

https://github.com/jrouwe/JoltPhysics/blob/de0b2e9eccbb67d47edb54677c551c195fcfd71a/Jolt/Physics/PhysicsSettings.h#L83-L86

If you set that value to something less than 1 it will work.

See #1406

**Comment by wedesoft** [comment]
Oh ok. I wasn't aware of this setting. That explains the behaviour :laughing: .
Thanks for your quick response!
